### DISCUSSION TITLE
Using Clip property to improve community's Classic GroupBox style for HeaderedContentControl

### DISCUSSION THREAD

**Discussion by BinToss** [AvaloniaUI/Avalonia#11902] [General]
I'll update [my reply in #823](https://github.com/AvaloniaUI/Avalonia/issues/823#issuecomment-1606959548) with the final result. I'm aware several community-created "GroupBox" controls exists, but the projects I could find did not support Avalonia 11.0.0 Previews or Release Candidate 1.1.

# 2024-03-18 - Release v1.0.0

https://github.com/BinToss/GroupBox.Avalonia/releases/v1.0.0
Thanks to @yll690 for the MultiBinding advice!
Thanks to @mgarstenauer for IMultiValueConverter help!

(2023-07-11) I'll come back to this at a later date. I broke and decided to use other solutions to port some of my projects to Avalonia.

<blockquote>

# 2023-07-11

![image](https://cdn.discordapp.com/attachments/569152356864753686/1128207412642451516/image.png)

Notes:
- Now targeting Avalonia 11.0.0
- Requires a Converter class to convert Bounds properties (Rect) to a Clip property (Geometry). 
  - When the Bounds are passed as Data Bindings with the Converter to a CombinedGeometry in AXAML, the CombinedGeomtry's Geometry1 and Geometry2 properties will observe the source Bounds properties for changes and will call the converter every time the Bounds change.
- If copying to your project, ensure you include the AXAML file in the project file e.g. `<AvaloniaResource Include="GUI\Styles\*.axaml" />`. Do not include AXAML files which have code-behind.
- This style's sample are made with `Avalonia.Themes.Simple`. When using the Simple theme, ensure it's added and untrimmed e.g. `<PackageReference Include="Avalonia.Themes.Simple" Version="11.0.0" />` and `<TrimmerRootAssembly Include="Avalonia.Themes.Simple" />`

Fixed:
- Header erroneously clipping Content has been resolved via workaround. The Border of GBContent is now invisible and not clipped. The visible, clipped border is now a separate element containing an invisible rectangle to ensure it has the same size as GBContent.

Known Issues:
- Resized Border does not render past its initial Bounds.
- Header text seems slightly offset on the Y axis.

TODO:
- Refactor to ThemeResource
- Get sample image for "classic" style with Fluent theme
- Get sample images for Simple- and Fluent- themed "modern" style.
- Consider making WPF-/MahApps.Metro-inspired theme

<details><summary>GroupBoxSimpleClassic.axaml</summary>

```xml
<Styles
    xmlns="https://github.com/avaloniaui"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    x:CompileBindings="True"
    xmlns:conv="clr-namespace:HXE.GUI.Converters;assembly=HXE">

    <Style Selector="HeaderedContentControl">

        <SimpleTheme />

        <Style.Resources>
            <conv:RectGeometryConverter x:Key="rectConv" />
            <Thickness x:Key="BorderMargin">0,10,0,0</Thickness>
            <Thickness x:Key="BorderPadding">5,10,5,5</Thickness>

        </Style.Resources>

        <Setter
            Property="Template">
            <Setter.Value>
                <ControlTemplate>
                    <Grid Name="GBGrid">
                        <Grid.RowDefinitions>
                            <RowDefinition Height="Auto" />
                            <RowDefinition Height="*" />
                        </Grid.RowDefinitions>
                        <Grid.ColumnDefinitions>
                            <ColumnDefinition Width="Auto" />
                            <ColumnDefinition Width="*" />
                        </Grid.ColumnDefinitions>

                        <TextBlock Name="GBHeaderText"
                            Margin="5,0,0,0"
                            Padding="5,0,5,0"
                            TextBlock.FontWeight="Bold" Text="{TemplateBinding Header}">
                        </TextBlock>

                        <Border Name="GBBorder"
                            Grid.RowSpan="2" Grid.ColumnSpan="2"
                            Margin="{DynamicResource BorderMargin}"
                            Padding="{DynamicResource BorderPadding}"
                            BorderBrush="{TemplateBinding Foreground}"
                            BorderThickness="1">
                            <Rectangle
                                Height="{Binding #GBContent.Height}"
                                Width="{Binding #GBContent.Width}" />
                            <Border.Clip>
                                <CombinedGeometry
                                    GeometryCombineMode="Exclude"
                                    Geometry1="{Binding #GBGrid.Bounds, Mode=OneWay, Converter={StaticResource rectConv}}"
                                    Geometry2="{Binding #GBHeaderText.Bounds, Mode=OneWay, Converter={StaticResource rectConv}}" />
                            </Border.Clip>
                        </Border>

                        <Border Name="GBContent"
                            Grid.RowSpan="2" Grid.ColumnSpan="2"
                            Margin="{DynamicResource BorderMargin}"
                            Padding="{DynamicResource BorderPadding}"
                            BorderThickness="0">

                            <ContentPresenter Content="{TemplateBinding Content}" />
                        </Border>
                    </Grid>
                </ControlTemplate>
            </Setter.Value>
        </Setter>
    </Style>
</Styles>
```

</details>
<details><summary>RectGeometryConverter.cs</summary>

```cs
using System;
using System.Globalization;
using Avalonia;
using Avalonia.Data;
using Avalonia.Data.Converters;
using Avalonia.Media;

namespace HXE.GUI.Converters;

public class RectGeometryConverter : IValueConverter
{
    public static readonly RectGeometryConverter Instance = new();

    public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        if (value is Rect rect)
            return new RectangleGeometry(rect);

        // converter used for the wrong type
        return new BindingNotification(new InvalidCastException(), BindingErrorType.Error);
    }

    public object ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        if (value is RectangleGeometry rectangle)
            return rectangle.Rect;

        // converter used for the wrong type
        return new BindingNotification(new InvalidCastException(), BindingErrorType.Error);
    }

}
```
</details>

----

# 2023-06-26

<details>

@derekantrican's examples are great, but the Classic style in that example will not adapt to Theme or Style changes i.e. Header background will remain solid white when the "real" background changes.
I've considered two solutions to this. One utlizes the Border's Clip property. The other utilizes the Border's OpacityMask. I'm attempting the former, but I'm struggling to make the Clip geometry recalculate when the source Bounds change. Both will likely require classes for conversions and/or observers.

## `Border.Clip = BoundsA - BoundsB`
<br/>
<details>
<summary>
<b>(Click to see images)</b> At one point, I was excluding <code>GBHeader.Bounds</code> (GBHeader being the Border containing the text block in the original example) from <code>GBContent.Bounds</code>, but an unexpectedly large part of the latter's Border is clipped despite the resulting geometry not affecting that region according the Dev Tools.
</summary>
<img src="https://cdn.discordapp.com/attachments/569152356864753686/1121703098612731934/image.png" />
<img src="https://cdn.discordapp.com/attachments/569152356864753686/1120640813077119017/image.png" />
</details>
<details><summary><b>(Click to see images)</b> However, I now exclude <code>GBHeaderText.Bounds</code> from <code>GBGrid.Bounds</code> to avoid that particular issue.</summary>
<img src="https://cdn.discordapp.com/attachments/569152356864753686/1122408984083705946/image.png" />
</details>

**This implementation currently has two issues.**
1. The `CombinedGeometry` assigned to `GBContent.Clip` is not recalculated when the `Rect`-value Bounds it was created from have changed.
2. The `CombinedGeomtry` assigned to `GBContent.Clip` partially obscures the presented contents. I believe this could be remedied by getting the `TransformBounds`, but the lack of this property in my Avalonia 11.0.0-rc1.1 application's Dev Tools window indicates it no longer exists or may be inaccessible under certain conditions. The suspected issue is that two Bounds have different origin points relative to the window, so excluding one Bounds from the other will result in one being offset from its expected position in the CombinedGeometry

<details><summary><h4>GroupBoxClassic.axaml</h4></summary>

```xml
<Styles
	xmlns="https://github.com/avaloniaui"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="https://github.com/avaloniaui ../../../AvaloniaSchema-11.0.0.xsd"
	xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
	x:CompileBindings="True"
	xmlns:conv="clr-namespace:HXE.GUI.Converters;assembly=HXE">

	<!-- AvaloniaSchema is generated by Avant Garde -->

	<Style
		Selector="HeaderedContentControl">
		<SimpleTheme />
		<Style.Resources>
			<conv:RectGeometryConverter x:Key="rectConv" />
		</Style.Resources>
		<Setter
			Property="Template">
			<Setter.Value>
				<ControlTemplate>
					<Grid Name="GBGrid">
						<Grid.RowDefinitions>
							<RowDefinition Height="Auto" />
							<RowDefinition Height="*" />
						</Grid.RowDefinitions>
						<Grid.ColumnDefinitions>
							<ColumnDefinition Width="Auto" />
							<ColumnDefinition Width="*" />
						</Grid.ColumnDefinitions>


						<TextBlock Name="GBHeaderText"
							Margin="5,0,0,0"
							Padding="5,0,5,0"
							ZIndex="1"
							TextBlock.FontWeight="Bold" Text="{TemplateBinding Header}">
							<TextBlock.Bounds />
						</TextBlock>

						<Border Name="GBContent"
							Grid.RowSpan="2" Grid.ColumnSpan="2"
							Margin="0,10,0,0"
							Padding="5,10,5,5"
							BorderBrush="{TemplateBinding Foreground}"
							BorderThickness="1"
							Clip="{Binding}">
							<Border.Clip>
								<CombinedGeometry x:DataType="Rect"
									GeometryCombineMode="Exclude"
									Geometry1="{Binding ElementName=GBGrid, Path=Bounds, Mode=OneWay, Converter={StaticResource rectConv}}"
									Geometry2="{Binding ElementName=GBHeaderText, Path=Bounds, Mode=OneWay, Converter={StaticResource rectConv}}">
								</CombinedGeometry>
							</Border.Clip>
							<ContentPresenter Content="{TemplateBinding Content}" />
						</Border>
					</Grid>
				</ControlTemplate>
			</Setter.Value>
		</Setter>
	</Style>
</Styles>

```
</details>
<details><summary><h4>RectGeometryConverter.cs</h4></summary>

```cs
using System;
using System.Globalization;
using Avalonia;
using Avalonia.Data;
using Avalonia.Data.Converters;
using Avalonia.Media;

namespace HXE.GUI.Converters;

public class RectGeometryConverter : IValueConverter
{
    public static readonly RectGeometryConverter Instance = new();

    public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        if (value is Rect rect)
            return new RectangleGeometry(rect);

        // converter used for the wrong type
        return new BindingNotification(new InvalidCastException(), BindingErrorType.Error);
    }

    public object ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        if (value is RectangleGeometry rectangle)
            return rectangle.Rect;

        // converter used for the wrong type
        return new BindingNotification(new InvalidCastException(), BindingErrorType.Error);
    }

}

```
</details>
</details>

</blockquote>

**Comment by mgarstenauer** [comment]
Here is my GroupBox implementation:

```cs
public sealed class GroupBox : HeaderedContentControl
{
}
```

```xaml
  <ControlTheme x:Key="{x:Type su:GroupBox}" TargetType="su:GroupBox">
    <Setter Property="BorderBrush" Value="Gray" />
    <Setter Property="BorderThickness" Value="1" />
    <Setter Property="CornerRadius" Value="3" />
    <Setter Property="Template">
      <ControlTemplate>
        <Grid ColumnDefinitions="6,Auto,*,6" RowDefinitions="Auto,Auto,*,6">
          <Border Name="Border"
                  Grid.Row="1"
                  Grid.RowSpan="3"
                  Grid.Column="0"
                  Grid.ColumnSpan="4"
                  Background="{TemplateBinding Background}"
                  BorderBrush="{TemplateBinding BorderBrush}"
                  BorderThickness="{TemplateBinding BorderThickness}"
                  CornerRadius="{TemplateBinding CornerRadius}">
            <Border.Clip>
              <MultiBinding Converter="{x:Static su:GapClipMaskConverter.Instance}">
                <Binding ElementName="Border" Path="Bounds" />
                <Binding ElementName="Header" Path="Bounds" />
              </MultiBinding>
            </Border.Clip>
          </Border>
          <ContentPresenter x:Name="Header"
                            Grid.Row="0"
                            Grid.RowSpan="2"
                            Grid.Column="1"
                            Padding="3,1,3,0"
                            Background="{TemplateBinding Background}"
                            Content="{TemplateBinding Header}"
                            ContentTemplate="{TemplateBinding HeaderTemplate}"
                            RecognizesAccessKey="True" />
          <ContentPresenter Name="PART_ContentPresenter"
                            Grid.Row="2"
                            Grid.Column="1"
                            Grid.ColumnSpan="2"
                            Padding="{TemplateBinding Padding}"
                            HorizontalContentAlignment="{TemplateBinding HorizontalContentAlignment}"
                            VerticalContentAlignment="{TemplateBinding VerticalContentAlignment}"
                            Content="{TemplateBinding Content}"
                            ContentTemplate="{TemplateBinding ContentTemplate}" />
        </Grid>
      </ControlTemplate>
    </Setter>
  </ControlTheme>
```

```cs
public sealed class GapClipMaskConverter : IMultiValueConverter
{
    public static readonly GapClipMaskConverter Instance = new();

    public object Convert(IList<object?> values, Type targetType, object? parameter,
                          CultureInfo culture)
    {
        if (values is not [Rect bounds, Rect gap]
            || bounds == default
            || gap == default)
        {
            return new BindingNotification(
                new ArgumentException(
                    "GapClipMaskConverter expects two non-empty rectangles (type Avalonia.Rect)."),
                BindingErrorType.Error);
        }

        gap = bounds.Intersect(gap);

        return new CombinedGeometry(
            GeometryCombineMode.Exclude,
            new RectangleGeometry { Rect = new(bounds.Size) },
            new RectangleGeometry { Rect = new(gap.Position - bounds.Position, gap.Size) });
    }
}
```

Essentially, I am using a `IMultiValueConverter` that always produces a new `CombinedGeometry` when the bounds change.
Hope that helps.

**Reply by BinToss** [reply]
Instead of recreating a Geometry every time the bounds change, I learned that I could instead create a single CombinedGeometry and bind its Geometry1 and Geometry2 to the bounds, utilizing a Rect-to-RectangleGeometry converter. The converter is called every time the values of the bindings update.

I recently found a workaround for my clipping issue, but discovered a new issue. My border resizes properly to anything smaller than its initial size, but anything outside of the initial size is clipped...despite the Clip property's image preview indicating the clipped region should be visible.
![image](https://github.com/AvaloniaUI/Avalonia/assets/7243190/380acd99-7f40-484b-8afc-426b88ad26bb)

----

I'll have to come back to this at a later date. I decided to use a more modern-styled GroupBox-like for the projects I'm porting to Avalonia.

**Comment by thevortexcloud** [comment]
FYI HeaderedContentControl now has a proper default style in 11.1. It wont look the same as yours though, but they seem fairly similar based on what I am seeing in the XAML for both.

https://github.com/AvaloniaUI/Avalonia/blob/1c792b294f82b26396e4f91032a2b535ea9c0669/src/Avalonia.Themes.Fluent/Controls/HeaderedContentControl.xaml

**Reply by BinToss** [reply]
Ah, yeah. I saw that while looking for themes' resources. It's nice!

**Reply by thevortexcloud** [reply]
Thank you, [I copied it from UWP](https://github.com/CommunityToolkit/WindowsCommunityToolkit/blob/6bfe8a5d21feb8818a8048b9fc41a84a70af8b0f/Microsoft.Toolkit.Uwp.UI.Controls.Layout/HeaderedContentControl/HeaderedContentControl.xaml#L5). Except I added a border since the UWP style does not have one.
